File: 6.controllers-services-repositories-middlewares-and-request-context/code/typescript/06the_service_layer_3.ts
export class BookService {
    constructor(private repo: any, private mailer: any) {}

    // No request, no response, no status codes ,  just logic.
    async listBooks(sort: string) {
        // Orchestration: ask the repository for what it needs.
        const books = await this.repo.findAllBooks(sort);
        // Could merge other repo calls, enrich, notify, etc.
        return books;
    }

    // A purely-logic service that never touches the DB:
    async notifyOwner(email: string) {
        return this.mailer.send(email, "Your book was added");
    }
}

File: 6.controllers-services-repositories-middlewares-and-request-context/code/rust/06the_service_layer_3.rs
pub struct BookService {
    repo: Box<dyn BookRepoTrait>,
    mailer: Box<dyn MailerTrait>,
}

impl BookService {
    // Notice: no request, no response, no status codes.
    // You cannot tell from this signature that it serves an API.
    pub async fn list_books(&self, sort: &str) -> Result<Vec<Book>, Box<dyn std::error::Error>> {
        // Orchestration: call the repository for the data it needs.
        let books = self.repo.find_all_books(sort).await?;
        // Could also: enrich, merge other repo calls, send notifications...
        Ok(books)
    }

    // A service that needs no database at all is perfectly valid:
    pub async fn notify_owner(&self, email: &str) -> Result<(), Box<dyn std::error::Error>> {
        self.mailer.send(email, "Your book was added").await
    }
}

// Sketches
pub trait BookRepoTrait {
    fn find_all_books(&self, sort: &str) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Vec<Book>, Box<dyn std::error::Error>>>>>;
}
pub trait MailerTrait {
    fn send(&self, email: &str, msg: &str) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<(), Box<dyn std::error::Error>>>>>;
}
pub struct Book {}
